src/services/classifier2020.ts

```diff
@@ -0,0 +1,257 @@
+                        if (node.parent) {
+                            const parentIsDeclaration = (isBindingElement(node.parent) || tokenFromDeclarationMapping[node.parent.kind] === typeIdx);
+                            if (parentIsDeclaration && (<NamedDeclaration>node.parent).name === node) {
+                                modifierSet = 1 << TokenModifier.declaration;
+                            }
+                        }
+
+                        // property declaration in constructor
+                        if (typeIdx === TokenType.parameter && isRightSideOfQualifiedNameOrPropertyAccess(node)) {
+                            typeIdx = TokenType.property;
+                        }
+
+                        typeIdx = reclassifyByType(typeChecker, node, typeIdx);
+
+                        const decl = symbol.valueDeclaration;
+                        if (decl) {
+                            const modifiers = getCombinedModifierFlags(decl);
+                            const nodeFlags = getCombinedNodeFlags(decl);
+                            if (modifiers & ModifierFlags.Static) {
+                                modifierSet |= 1 << TokenModifier.static;
+                            }
+                            if (modifiers & ModifierFlags.Async) {
+                                modifierSet |= 1 << TokenModifier.async;
+                            }
+                            if (typeIdx !== TokenType.class && typeIdx !== TokenType.interface) {
+                                if ((modifiers & ModifierFlags.Readonly) || (nodeFlags & NodeFlags.Const) || (symbol.getFlags() & SymbolFlags.EnumMember)) {
+                                    modifierSet |= 1 << TokenModifier.readonly;
+                                }
+                            }
+                            if ((typeIdx === TokenType.variable || typeIdx === TokenType.function) && isLocalDeclaration(decl, sourceFile)) {
+                                modifierSet |= 1 << TokenModifier.local;
+                            }
+                            if (program.isSourceFileDefaultLibrary(decl.getSourceFile())) {
+                                modifierSet |= 1 << TokenModifier.defaultLibrary;
+                            }
+                        }
+                        else if (symbol.declarations && symbol.declarations.some(d => program.isSourceFileDefaultLibrary(d.getSourceFile()))) {
+                            modifierSet |= 1 << TokenModifier.defaultLibrary;
+                        }
+
+                        collector(node, typeIdx, modifierSet);
+
+                    }
+                }
+            }
+            forEachChild(node, visit);
+
+            inJSXElement = prevInJSXElement;
+        }
+        visit(sourceFile);
+    }
+
+    function classifySymbol(symbol: Symbol, meaning: SemanticMeaning): TokenType | undefined {
+        const flags = symbol.getFlags();
+        if (flags & SymbolFlags.Class) {

```

Comment:
Do you need early return if 
```ts
 if ((flags & SymbolFlags.Classifiable) === SymbolFlags.None) 
```

Reply:
It could, but I'd need to make a new SymbolFlag for it as there are things like functions and variable declarations in this classifier also, do you have an opinion either way?

Reply:
No but on that note. Is this more additive compared to our original classifier? If so why not use this all the time and filter out things in original format mode ?

Reply:
The results sent back are different, and conform to a WIP LSP spec. The current one is used by VS, so we basically need to support both. I think [this test](https://github.com/microsoft/TypeScript/pull/39119/files#diff-b480226546c2e6363be7c39e5bcdba4aL19) shows the differences in the results quite well:

```ts
var c = classification;
const c = classification("original");
verify.syntacticClassificationsAre(
     c.comment(firstCommentText),
     c.keyword("function"), c.identifier("myFunction"), c.punctuation("("), c.comment("/* x */"), c.parameterName("x"), c.punctuation(":"), c.keyword("any"), c.punctuation(")"), c.punctuation("{"),
     c.keyword("var"), c.identifier("y"), c.operator("="), c.identifier("x"), c.operator("?"), c.identifier("x"), c.operator("++"), c.operator(":"), c.operator("++"), c.identifier("x"), c.punctuation(";"),
     c.punctuation("}"),
     c.comment("// end of file")); 
     c.comment("// end of file"));

const c2 = classification("2020");
 verify.semanticClassificationsAre("2020",
     c2.semanticToken("function.declaration", "myFunction"), 
     c2.semanticToken("parameter.declaration", "x"), 
     c2.semanticToken("variable.declaration.local", "y"), 
     c2.semanticToken("parameter", "x"), 
     c2.semanticToken("parameter", "x"), 
     c2.semanticToken("parameter", "x"), 
 );
```

The first format (original) includes everything form comments to braces, the second (LSP) is quite an explicit a subset of those results

Reply:
But if thats the case isnt it better to classify things in one place as function. variable, comments etc and filter out and swithc things depending on format ?

Eg. returned "function.declaration" becomes identifier in original. 
punctuation, comment is ignored in 2020 etc

Reply:
That not an unreasonable idea! However the current implementation is a scanner, but I need to use an AST to get resolved information for things which are outside the scope of the current file  

The subsequent PR from this https://github.com/orta/TypeScript/pull/3/commits/58e830c3d6a5bbceaecd599ebad5255e0cde5e62#diff-830691d11bb3d3f0f7ca90ef0ee364afR266 speeds up the processing but requires the AST ahead of time, and it could be implemented on top of the current classifier scanner - there's quite a mis-match between what they're doing

Reply:
Poke on ^ - is that reasonable?